src/c1_state_machine/p4_accounted_currency.rs

```diff
@@ -43,4 +44,268 @@ impl StateMachine for AccountedCurrency {
+    let start = HashMap::from([(User::Alice, 100)]);
+    let end = AccountedCurrency::next_state(
+        &start,
+        &AccountingTransaction::Burn {
+            burner: User::Bob,
+            amount: 50,
+        },
+    );
+    let expected = HashMap::from([(User::Alice, 100)]);
+
+    assert_eq!(end, expected);
+}
+
+#[test]
+fn sm_4_simple_transfer() {
+    let start = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+    let end = AccountedCurrency::next_state(
+        &start,
+        &AccountingTransaction::Transfer {
+            sender: User::Alice,
+            receiver: User::Bob,
+            amount: 10,
+        },
+    );
+    let expected = HashMap::from([(User::Alice, 90), (User::Bob, 60)]);
+
+    assert_eq!(end, expected);
+
+    let start = HashMap::from([(User::Alice, 90), (User::Bob, 60)]);
+    let end1 = AccountedCurrency::next_state(
+        &start,
+        &AccountingTransaction::Transfer {
+            sender: User::Bob,
+            receiver: User::Alice,
+            amount: 50,
+        },
+    );
+    let expected1 = HashMap::from([(User::Alice, 140), (User::Bob, 10)]);
+
+    assert_eq!(end1, expected1);
+}
+
+#[test]
+fn sm_4_send_to_same_user() {
+    let start = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+    let end = AccountedCurrency::next_state(
+        &start,
+        &AccountingTransaction::Transfer {
+            sender: User::Bob,
+            receiver: User::Bob,
+            amount: 10,
+        },
+    );
+    let expected = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+
+    assert_eq!(end, expected);
+}
+
+#[test]
+fn sm_4_unsufficient_balance_transfer() {
+    let start = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+    let end = AccountedCurrency::next_state(
+        &start,
+        &AccountingTransaction::Transfer {
+            sender: User::Bob,
+            receiver: User::Alice,
+            amount: 60,
+        },
+    );
+    let expected = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+
+    assert_eq!(end, expected);
+}
+
+#[test]
+fn sm_4_sender_not_registered() {
+    let start = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+    let end = AccountedCurrency::next_state(
+        &start,
+        &AccountingTransaction::Transfer {
+            sender: User::Charlie,
+            receiver: User::Alice,
+            amount: 50,
+        },
+    );
+    let expected = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+
+    assert_eq!(end, expected);
+}
+
+#[test]
+fn sm_4_receiver_not_registered() {
+    let start = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+    let end = AccountedCurrency::next_state(
+        &start,
+        &AccountingTransaction::Transfer {
+            sender: User::Alice,
+            receiver: User::Charlie,
+            amount: 50,
+        },
+    );
+    let expected = HashMap::from([(User::Alice, 100), (User::Bob, 50)]);
+
+    assert_eq!(end, expected);
+}
```

Comment:
This should create the account by adding an entry for charlie into the map. This is the main way accounts get created since in a more realistic state machine, arbitrary users can't mint.